File: day01/answer.js
import fs from 'fs/promises';

const [inputPath, debugStr] = process.argv.slice(2);

const debug = /debug=true/.test(debugStr);
function findDepthChanges(input) {
  let increased = 0;
  let decreased = 0;
  let stayed = 0;
  for (let i = 1; i < input.length; i++) {
    if (input[i] < input[i - 1]) {
      debug && console.log(`${input[i - 1]} -> ${input[i]} (decreased)`);
      decreased++;
    } else if (input[i] > input[i - 1]) {
      debug && console.log(`${input[i - 1]} -> ${input[i]} (increased)`);
      increased++;
    } else {
      debug && console.log(`${input[i - 1]} -> ${input[i]} (stayed)`);
      stayed++;
    }
  }
  console.log(`
Individual Results (${input.length} total):
  Increased: ${increased}
  Decreased: ${decreased}
  Stayed: ${stayed}`);
}

function findGroupedDepthChanges(input) {
  let increased = 0;
  let decreased = 0;
  let stayed = 0;
  for (let i = 3; i < input.length; i++) {
    const groupA = [input[i - 1], input[i - 2], input[i - 3]];
    const groupB = [input[i], input[i - 1], input[i - 2]];
    const groupASum = groupA.reduce(sum, 0);
    const groupBSum = groupB.reduce(sum, 0);
    if (groupBSum < groupASum) {
      debug && console.log(`${groupA} -> ${groupB} (decreased)`);
      decreased++;
    } else if (groupBSum > groupASum) {
      debug && console.log(`${groupA} -> ${groupB} (increased)`);
      increased++;
    } else {
      debug && console.log(`${groupA} -> ${groupB} (stayed)`);
      stayed++;
    }
  }
  console.log(`
Group Results (${input.length} total):
  Increased: ${increased}
  Decreased: ${decreased}
  Stayed: ${stayed}`);
}

function sum(total, current) {
  return total + current;
}

(async () => {
  const input = (await fs.readFile(inputPath, 'utf-8'))
    .trim()
    .split('\n')
    .map((l) => parseInt(l));

  // part 1
  findDepthChanges(input);

  // part 2
  findGroupedDepthChanges(input);
})();

File: day01/answer.rs
use std::{fs, env};

fn main() -> () {
    let input_path: String = match env::args().nth(1) {
        Some(p) => p,
        None => {
            println!("no input path");
            return;
        },
    };
    let file: String = match fs::read_to_string(input_path) {
        Ok(p) => p,
        Err(e) => {
            println!("could not read input file: {}", e);
            return;
        },
    };

    let values = file.lines().map(|line| line.parse::<i32>().unwrap());
    let skipped = values.clone().skip(1);
    let zipped = values.clone().zip(skipped);
    let mut total = 0;
    let mut increased = 0;
    let mut decreased = 0;
    let mut stayed = 0;
    for (prev, curr) in zipped {
        total += 1;
        if prev < curr {
            increased += 1;
        } else if prev > curr {
            decreased += 1;
        } else {
            stayed += 1;
        }
    }
    println!("Results ({}):\n\tIncreased: {}\n\tDecreased: {}\n\tStayed: {}", total, increased, decreased, stayed);

    let vals: Vec<i32> = values.collect();
    let total = vals.len();
    let mut increased = 0;
    let mut decreased = 0;
    let mut stayed = 0;
    for i in 3..vals.len() {
        let group_a = vals[i - 3] + vals[i - 2] + vals[i - 1];
        let group_b = vals[i - 2] + vals[i - 1] + vals[i];
        if group_a < group_b {
            increased += 1;
        } else if group_a > group_b {
            decreased += 1;
        } else {
            stayed += 1;
        }
    }
    println!("Group results ({}):\n\tIncreased: {}\n\tDecreased: {}\n\tStayed: {}", total, increased, decreased, stayed);
}
